### PR TITLE
p2p: fix IPv4 subnet arithmetic

### LINKED ISSUE #10325: Ban list should implement subnet set arithmetic
`monerod`’s current ban list implementation does not properly handle IP subnets.

## Problem

- Banning a larger subnet (e.g. `172.16.0.0/12`) and then attempting to unban with a smaller subnet (e.g. `172.16.1.0/24`) does not correctly remove the larger subnet from the list.
- Unbanning all IPs via `monerod unban 0.0.0.0/0` does not work.
- No easy way to unban all IPs

Example of the issue:
```bash
# Ban 172.16.0.0/12 subnet
$ monerod ban 172.16.0.0/12
Monero 'Fluorine Fermi' (v0.18.4.5-release)

# Unban a subset of the above subnet
$ monerod unban 172.16.1.0/24
Monero 'Fluorine Fermi' (v0.18.4.5-release)

# Validate unban
$ monerod banned 172.16.1.1
Monero 'Fluorine Fermi' (v0.18.4.5-release)
172.16.1.1 is banned for 86359 seconds # Still banned
```

## Expected behavior
Subnet ban/unban operations should respect proper subnet set arithmetic:
- If a broad subnet (e.g. `172.16.0.0/12`) is banned, it should implicitly ban all subnets within it.
- If a subnet is unbanned that overlaps a previously banned larger subnet, the implementation should correctly recompute the resulting ban set.

### PR DESCRIPTION
**Pull request by Ap4sh**
Fixes #10325

This fixes two related IPv4 subnet banlist cases:

- non-octet CIDR prefixes like `172.16.0.0/12` are now masked using the correct bit layout
- unbanning a subnet inside a broader blocked subnet now splits the remaining blocked range instead of leaving the broader ban unchanged

Explicit IPv4 host bans covered by the unbanned subnet are removed too, so `unban 0.0.0.0/0` clears the IPv4 banlist as expected

Tests:

- `cmake --build build/local-output-selection-tests --target unit_tests -j4`
- `build/local-output-selection-tests/tests/unit_tests/unit_tests --gtest_filter=ban.*:node_server.*:get_network_address.ipv4subnet`
- `build/local-output-selection-tests/tests/unit_tests/unit_tests --gtest_filter=host.*:tor_address.*:get_network_address.*:i2p_address.*:get_network_address_host_and_port.*:socks_endpoint.*:socks_client.*:socks_connector.*`

### CODE REVIEW

**Comment by selsta** [comment]
Please undo unrelated whitespace changes

**Comment by Ap4sh** [comment]
fixed, thanks

File: tests/unit_tests/node_server.cpp [hunk 1]
```diff
@@ -345,14 +341,64 @@ TEST(ban, subnet)
...
+  ASSERT_TRUE(server.unblock_subnet(MAKE_IPV4_SUBNET(1,2,3,8,24)));
+  ASSERT_TRUE(server.get_blocked_subnets().size() == 16);
+  ASSERT_FALSE(server.is_host_blocked(MAKE_IPV4_ADDRESS(1,2,3,8), &seconds));
+  ASSERT_TRUE(server.is_host_blocked(MAKE_IPV4_ADDRESS(1,2,2,255), &seconds));
+  ASSERT_TRUE(server.is_host_blocked(MAKE_IPV4_ADDRESS(1,2,4,0), &seconds));
   ASSERT_TRUE(server.unblock_subnet(MAKE_IPV4_SUBNET(1,255,0,0,8)));
   ASSERT_TRUE(server.get_blocked_subnets().size() == 0);
 }
 
+TEST(ban, subnet_arithmetic)
```
Comment (selsta) [inline tests/unit_tests/node_server.cpp:353]: Could you add explicit `/0` and `/32` boundary cases?

File: src/p2p/net_node.inl [hunk 2]
```diff
@@ -77,6 +77,65 @@ static inline boost::asio::ip::address_v4 make_address_v4_from_v6(const boost::a
 
 namespace nodetool
 {
+  namespace
+  {
+    uint64_t ipv4_subnet_size(const uint8_t mask)
+    {
+      return mask == 0 ? (uint64_t{1} << 32) : (uint64_t{1} << (32 - mask));
```
Comment (selsta) [inline src/p2p/net_node.inl:85]:
```suggestion
      CHECK_AND_ASSERT_THROW_MES(mask <= 32, "invalid IPv4 subnet mask");
      return mask == 0 ? (uint64_t{1} << 32) : (uint64_t{1} << (32 - mask));
```

does this make sense here?

File: tests/unit_tests/node_server.cpp (hunk 1 above)
Comment (Ap4sh) [reply tests/unit_tests/node_server.cpp:353]: added explicit `/32` and `/0` coverage now, the `/0` case also removes a single `/32` from the full range so the split path gets checked too

File: src/p2p/net_node.inl (hunk 2 above)
Comment (Ap4sh) [reply src/p2p/net_node.inl:85]: yeah, makes sense, added the `mask <= 32` guard before the shift

File: contrib/epee/src/net_utils_base.cpp [hunk 3]
```diff
@@ -17,6 +17,15 @@ static inline uint32_t make_address_v4_from_v6(const boost::asio::ip::address_v6
   return htonl(v4);
 }
 
+static inline uint32_t make_ipv4_cidr_mask(const uint8_t mask)
+{
+  if (mask == 0)
+    return 0;
+  if (mask > 32)
```
Comment (selsta) [inline contrib/epee/src/net_utils_base.cpp]: Since the other place now throws when > any way to make this more strict instead of silently treating as /32? Maybe in the constructor?

Comment (Ap4sh) [reply contrib/epee/src/net_utils_base.cpp]: done, made the subnet constructor reject masks above 32 now

i left the cidr mask helper non-throwing since `subnet()` and the comparison path are still `noexcept`, and added a `/33` constructor check. local subnet tests pass too

File: contrib/epee/src/net_utils_base.cpp [hunk 4]
```diff
@@ -17,6 +17,15 @@ static inline uint32_t make_address_v4_from_v6(const boost::asio::ip::address_v6
...
+    return 0xffffffffu;
```
Comment (selsta) [inline contrib/epee/src/net_utils_base.cpp]: This branch can be removed now, it's unreachable.

File: src/p2p/net_node.inl [hunk 5]
```diff
@@ -379,10 +450,35 @@ namespace nodetool
...
+        unblocked = true;
+      for (const auto &entry : remaining)
+        emplace_blocked_subnet(blocked_subnets, entry, blocked_subnet.second);
+    }
+
+    if (!unblocked)
       return false;
-    m_blocked_subnets.erase(i);
+    m_blocked_subnets.swap(blocked_subnets);
     MCLOG_CYAN(el::Level::Info, "global", "Subnet " << subnet.host_str() << " unblocked.");
```
Comment (selsta) [inline src/p2p/net_node.inl]:
```suggestion
    MCLOG_CYAN(el::Level::Info, "global", "Subnet " << subnet.str() << " unblocked.");
```

Currently this is equivalent but this is a bit clearer since it logs a subnet object and not a host.

File: src/p2p/net_node.inl [hunk 6]
```diff
@@ -379,10 +450,35 @@ namespace nodetool
...
+        unblocked = true;
+      }
+      else
+        ++i;
+    }
+
+    std::map<epee::net_utils::ipv4_network_subnet, time_t> blocked_subnets;
+    for (const auto &blocked_subnet : m_blocked_subnets)
+    {
+      std::vector<epee::net_utils::ipv4_network_subnet> remaining;
```
Comment (selsta) [inline src/p2p/net_node.inl]: Small easy cleanup: `remaining` can be moved outside the loop and reused with `clear()`, optionally with `reserve(32)`.

**Review by selsta — COMMENTED** [review]
Please also add a test case for the following

```
ban 172.16.0.0/16
ban 172.16.1.0/24
unban 172.16.0.0/16
```

expected result

```
172.16.1.1 is not blocked
blocked_subnets is empty
```

This was specifically mentioned in https://github.com/monero-project/monero/pull/10324#issuecomment-3905421983.

**Comment by Ap4sh** [comment]
done, added that nested `/16` + `/24` case and cleaned up the three review notes too

i rebuilt `unit_tests` and reran `ban.subnet_arithmetic` locally

**Review by selsta — APPROVED** [review]

**Review by jpk68 — APPROVED** [review]
LGTM. Thanks for fixing this :)
